### PR TITLE
In-memory engine: fix seek does not set self.valid properly

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Ref #16141

What's Changed:

```commit-message
fix seek does not set self.valid properly
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/region_cache_memory_engine/src/read.rs [hunk 1]
```diff
@@ -797,6 +799,42 @@ mod tests {
...
+        }
+
+        let snapshot = engine.snapshot(range.clone(), u64::MAX, 100).unwrap();
+        let mut iter_opt = IterOptions::default();
+        iter_opt.set_upper_bound(&range.end, 0);
+        iter_opt.set_lower_bound(&range.start, 0);
+        let mut iter = snapshot.iterator_opt("write", iter_opt.clone()).unwrap();
+
+        let key = construct_mvcc_key("b", 10);
+        iter.seek(&key).unwrap();
```
Comment (LykxSassinator) [inline components/region_cache_memory_engine/src/read.rs:826]:
```suggestion
        iter.seek(&key).unwrap();
        assert!(iter.valid().unwrap();
```

File: components/region_cache_memory_engine/src/read.rs [hunk 2]
```diff
@@ -797,6 +799,42 @@ mod tests {
...
+        let key = construct_mvcc_key("d", 10);
+        iter.seek(&key).unwrap();
+        assert!(!iter.valid().unwrap());
+
+        let key = construct_mvcc_key("b", 10);
+        iter.seek_for_prev(&key).unwrap();
```
Comment (LykxSassinator) [inline components/region_cache_memory_engine/src/read.rs:832]: ditto.

**Review by LykxSassinator — COMMENTED** [review]
rest LGTM

**Review by LykxSassinator — APPROVED** [review]

**Review by afeinberg — APPROVED** [review]

**Review by Connor1996 — APPROVED** [review]

File: components/region_cache_memory_engine/src/read.rs [hunk 3]
```diff
@@ -312,6 +312,8 @@ impl RangeCacheIterator {
         self.iter.seek(key, guard);
         if self.iter.valid() {
             self.find_next_visible_key(false, guard);
+        } else {
+            self.valid = false;
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/read.rs:316]: What about other places when self.iter.valid() is false and should we set the self.valid = false as well?

**Pull request merged**